src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{
+    chain::chain_information::{
+        babe_fetch_epoch, BabeEpochInformation, ChainInformation, ChainInformationConsensus,
+        ChainInformationFinality,
+    },
+    executor::{
+        host::{HostVmPrototype, NewErr},
+        vm::ExecHint,
+    },
+    finality::{grandpa::warp_sync, justification::verify},
+    header::Header,
+    libp2p::PeerId,
+    network::protocol::GrandpaWarpSyncResponseFragment,
+};
+
+/// Problem encountered during a call to [`grandpa_warp_sync`].
+#[derive(Debug, derive_more::Display)]
+pub enum Error {
+    #[display(fmt = "All requests failed.")]
+    AllRequestsFailed,
+    #[display(fmt = "{}", _0)]
+    Verifier(verify::Error),
+    #[display(fmt = "{}", _0)]
+    BabeFetchEpoch(babe_fetch_epoch::Error),
+    #[display(fmt = "{}", _0)]
+    NewRuntime(NewErr),
+}
+
+/// The configuration for [`grandpa_warp_sync`].
+pub struct Config<'a> {
+    /// A list of connected peers.
+    pub connected_peers: Vec<PeerId>,
+    /// The chain information of the genesis block.
+    pub genesis_chain_information: &'a ChainInformation,
+}
+
+/// Starts syncing via grandpa warp sync.
+pub fn grandpa_warp_sync(config: Config) -> GrandpaWarpSync {
+    GrandpaWarpSync::WarpSyncRequest(WarpSyncRequest {
+        peer_index: 0,
+        connected_peers: config.connected_peers,
+        genesis_chain_information: config.genesis_chain_information,
+    })
+}
+
+/// The grandpa warp sync state machine.
+pub enum GrandpaWarpSync<'a> {
```

Comment:
```suggestion
pub enum GrandpaWarpSync {
```

Please remove the lifetime everywhere. It makes things way too annoying to use.